src/crimson/osd/pg.cc

```diff
@@ -1089,37 +1085,34 @@ PG::do_osd_ops_execute(
      rollbacker, failure_func_ptr]
     (const std::error_code& e) mutable {
     ceph_tid_t rep_tid = shard_services.get_tid();
-    return rollbacker.rollback_obc_if_modified(e).then_interruptible(
-    [&, op_info, m, obc,
-     this, e, rep_tid, failure_func_ptr] {
-      // record error log
-      auto maybe_submit_error_log =
-        seastar::make_ready_future<std::optional<eversion_t>>(std::nullopt);
-      // call submit_error_log only for non-internal clients
-      if constexpr (!std::is_same_v<Ret, void>) {
-        if(op_info.may_write()) {
-          maybe_submit_error_log =
-            submit_error_log(m, op_info, obc, e, rep_tid);
-        }
+    rollbacker.rollback_obc_if_modified(e);
+    // record error log
+    auto maybe_submit_error_log =
+      seastar::make_ready_future<std::optional<eversion_t>>(std::nullopt);
+    // call submit_error_log only for non-internal clients
+    if constexpr (!std::is_same_v<Ret, void>) {
+      if(op_info.may_write()) {
+        maybe_submit_error_log =
+          submit_error_log(m, op_info, obc, e, rep_tid);
```

Comment:
https://pulpito.ceph.com/matan-2024-09-11_13:53:46-crimson-rados-wip-matanb-crimson-testing-sep11-distro-crimson-smithi/

Seems from testing that the obc may have not actually rollbacked here.
Both only failures are `TestClsRbd.directory_methods`.
We had already failed this test in the past and it was fixed here - https://github.com/ceph/ceph/pull/54931/commits/999845c25121cf8d15fdce6b23ee5b3c4b584a9b.

Reply:
Perhaps the `obc` submitted in l.1096 is not actually referencing the one that was rollbacked? 
Will look into this tomorrow.

Reply:
> Perhaps the `obc` submitted in l.1096 is not actually referencing the one that was rollbacked? Will look into this tomorrow.

Yes, I think so, the following change should fix the issue:
```
diff --git a/src/crimson/osd/pg.cc b/src/crimson/osd/pg.cc
index f8473daecca..d100775173c 100644
--- a/src/crimson/osd/pg.cc
+++ b/src/crimson/osd/pg.cc
@@ -1081,8 +1081,7 @@ PG::do_osd_ops_execute(
       std::move(all_completed_fut)
     );
   }, OpsExecuter::osd_op_errorator::all_same_way(
-    [this, op_info, m, obc,
-     rollbacker, failure_func_ptr]
+    [this, op_info, m, rollbacker, failure_func_ptr]
     (const std::error_code& e) mutable {
     ceph_tid_t rep_tid = shard_services.get_tid();
     rollbacker.rollback_obc_if_modified(e);
@@ -1093,7 +1092,7 @@ PG::do_osd_ops_execute(
     if constexpr (!std::is_same_v<Ret, void>) {
       if(op_info.may_write()) {
         maybe_submit_error_log =
-          submit_error_log(m, op_info, obc, e, rep_tid);
+          submit_error_log(m, op_info, rollbacker.get_obc(), e, rep_tid);
       }
     }
     return maybe_submit_error_log.then_interruptible(

```

Reply:
Just fixed the code, please test this PR again.